cloudevents/working-drafts/dds-format.md

```diff
@@ -0,0 +1,247 @@
+
+OPTIONAL and extension attributes are represented using a sequence of key-value
+constructs (DDS-CE-Attributes) enabling direct support of the CloudEvent
+[type system][ce-types]. An underlying union and enumeration data structure
+are needed to construct the key-value sequence to establish the association
+between the IDL types and the CloudEvent type system.
+
+```xml
+
+  <enum name="AttributeType">
+   <enumerator name="BOOL"/>
+   <enumerator name="INT32"/>
+   <enumerator name="STRING"/>
+   <enumerator name="BYTES"/>
+   <enumerator name="URI"/>
+   <enumerator name="URI_REF"/>
+   <enumerator name="TIMESTAMP"/>
+ </enum>
+	  
+ <union name="AttributeValue" extensibility="final">
+   <discriminator type="nonBasic" nonBasicTypeName="io::cloudevents::AttributeType"/>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::BOOL)"/>
+     <member name="ce_boolean" type="boolean"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::INT32)"/>
+     <member name="ce_integer" type="int32"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::STRING)"/>
+     <member name="ce_string" type="string" stringMaxLength="255"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::BYTES)"/>
+     <member name="ce_bytes" type="byte" sequenceMaxLength="100"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::URI)"/>
+     <member name="ce_uri" type="nonBasic" nonBasicTypeName="io::cloudevents::URI"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::URI_REF)"/>
+     <member name="ce_uri_reference" type="nonBasic" nonBasicTypeName="io::cloudevents::URI_Ref"/>
+   </case>
+ </union>
+
+ <struct name="DDS-CE-Attribute">
+   <member name="key" type="string" stringMaxLength="255"/>
+   <member name="value" type="nonBasic" nonBasicTypeName="io::cloudevents::AttributeValue"/>
+ </struct>
+
+ <typedef name="DDS-CE-Attributes" type="nonBasic" nonBasicTypeName="io::cloudevents::DDS-CE-Attribute" sequenceMaxLength="100"/>
+
+```
+In this model an attribute's name is used as the *key* and is associated
+with its *value* stored in the appropriately typed property.
+
+This approach allows attributes to be represented and transported
+with no loss of *type* information.
+
+## 3. Data
+
+The specification allows for data payloads of the following types to be explicitly represented:
+
+* binary
+* string
+
+```xml
+
+ <enum name="DataKind">
+   <enumerator name="BINARY"/>
+   <enumerator name="TEXT"/>
+ </enum>
+	  
+ <union name="Data" extensibility="final">
+   <discriminator type="nonBasic" nonBasicTypeName="io::cloudevents::DataKind"/>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::BINARY)"/>
+     <member name="binary_data" type="byte" sequenceMaxLength="100"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::TEXT)"/>
+     <member name="text_data" type="string" stringMaxLength="255"/>
```

Comment:
Similar to @duglin comment above ... I'm not sure we place any constraints (rightly or wrongly) on the size of event data.

Reply:
I removed all bounds on length/sizes, so we should be good here.